safe_prime_generator.go

```diff
@@ -0,0 +1,301 @@
+
+// The code is the original Go implementation of rand.Prime optimized for
+// generating safe (Sophie Germain) primes.
+// A safe prime is a prime number of the form 2p + 1, where p is also a prime.
+
+package paillier
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"sync"
+	"time"
+)
+
+// smallPrimes is a list of small, prime numbers that allows us to rapidly
+// exclude some fraction of composite candidates when searching for a random
+// prime. This list is truncated at the point where smallPrimesProduct exceeds
+// a uint64. It does not include two because we ensure that the candidates are
+// odd by construction.
+var smallPrimes = []uint8{
+	3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53,
+}
+
+// smallPrimesProduct is the product of the values in smallPrimes and allows us
+// to reduce a candidate prime by this number and then determine whether it's
+// coprime to all the elements of smallPrimes without further big.Int
+// operations.
+var smallPrimesProduct = new(big.Int).SetUint64(16294579238595022365)
+
+// GenerateSafePrime tries to find a safe prime concurrently.
+// The returned result is a safe prime `p` and prime `q` such that `p=2q+1`.
+// Concurrency level can be controlled with the `concurrencyLevel` parameter.
+// If a safe prime could not be found in the specified `timeout`, the error
+// is returned. Also, if at least one search process failed, error is returned
+// as well.
+//
+// How fast we generate a prime number is mostly a matter of luck and it depends
+// on how lucky we are with drawing the first bytes.
+// With today's multicore processors, we can execute the process on multiple
+// cores concurrently, accept the first valid result and cancel the rest of
+// work. This way, with the same finding algorithm, we can get the result
+// faster.
+//
+// Concurrency level should be set depending on what `bitLen` of prime is
+// expected. For example, as of today, on a typical workstation, for 512-bit
+// safe prime, `concurrencyLevel` should be set to `1` as generating the prime
+// of this length is a matter of milliseconds for a single core.
+// For 1024-bit safe prime, `concurrencyLevel` should be usually set to at least
+// `2` and for 2048-bit safe prime, `concurrencyLevel` must be set to at least
+// `4` to get the result in a reasonable time.
+//
+// This function generates safe primes of at least 6 `bitLen`. For every
+// generated safe prime, the two most significant bits are always set to `1`
+// - we don't want the generated number to be too small.
+func GenerateSafePrime(
+	bitLen int,
+	concurrencyLevel int,
+	timeout time.Duration,
+	random io.Reader,
+) (p *big.Int, q *big.Int, err error) {
+	if bitLen < 6 {
+		return nil, nil, errors.New("safe prime size must be at least 6 bits")
+	}
+
+	primeChan := make(chan safePrime, concurrencyLevel)
+	errChan := make(chan error, concurrencyLevel)
+
+	defer close(primeChan)
+	defer close(errChan)
+
+	mutex := &sync.Mutex{}
```

Comment:
I think you can get rid of this mutex. The context package is just constructors for the signal crossing API boundaries, and should be threadsafe in the way that you're using it. I'm not 100% on that, so I pulled your code, removed all references to `mutex`, and ran `go test -v -race -count=40` a few times just to be sure. All was fine.

Reply:
It sometimes fails without the mutex:
```
$ go test -count 100
panic: send on closed channel

goroutine 291 [running]:
github.com/keep-network/paillier.runGenPrimeRoutine.func1(0x1180e00, 0xc42001eb80, 0x11804c0, 0xc42008a2d0, 0xc4202523e4, 0x2, 0x2, 0xc4201c8420, 0x1, 0xc42040e980, ...)
        /Users/piotr/go/src/github.com/keep-network/paillier/safe_prime_generator.go:272 +0x483
created by github.com/keep-network/paillier.runGenPrimeRoutine
        /Users/piotr/go/src/github.com/keep-network/paillier/safe_prime_generator.go:179 +0x21c
exit status 2
FAIL    github.com/keep-network/paillier        1.082s
```

My theory is that worker thread is preempted after the `if ctx.Err() == nil ` check and someone cancels the context in a meantime. It results in closing the channel, and when thread has the control back, it can't `primeChan <- safePrime{p, q}`. 

However, this `mutex` was added before we got the `waitGroup.Wait()` in the main thread. My understanding how `waitGroup.Wait()`  and `defer close(primeChan)` works is that `close(primeChan)` should not be called before anything other in `GenerateSafePrime` including `waitGroup.Wait()`. So I don't really understand why is it possible that we closed the channel before we received `defer waitGroup.Done()` from all the workers and thus `waitGroup.Wait()` allowed to go further.

Reply:
My fault, instead of stacking changes I should try them one by one. The comment above is no longer valid. You are right, we can remove mutex and cancelled context check.